### DISCUSSION TITLE
Filesystem registry, overlay ports, or something else?

### DISCUSSION THREAD

**Discussion by thxkiwi** [microsoft/vcpkg#19784] [Q&A]
The scenario:

- Library; written in-house
  - Built using vcpkg (manifest mode) to manage libraries on which Library depends
  - Released versions reside in a vcpkg registry (kind: git)
- App using library; also written in-house
  - Built using vcpkg (manifest mode) to acquire Library and Library's dependencies

Sometimes App must be linked against a released version of Library. Sometimes App must be linked against a "local" built development version of Library.  However, having to push development changes of Library through a git based vcpkg registry adds a lot of extra work to the development process for a developer to test before committing code.

So, when building App, I've been using a vcpkg registry (kind: filesystem) to acquire and build development versions of modified Library source. This is a little better but with binary caching one must bump something (I've been using port-version) in the filesystem based vcpkg registry to force a rebuild of Library. Still not ideal.

I was just considering `--overlap-ports` but I'm not sure I'm heading down the right path with that nor am I sure how that would work with manifest mode.

Is there a recommended way to build App and Library (using vcpkg to manage the dependencies of Library) that allows switching between a vcpkg registry of released versions versus some way of building development versions of Library and linking App against those?

**Comment by autoantwort** [comment]
Maybe add an option that switches between using vcpkg (for releases) and FetchContent (for dev builds)(or find_package when it is already installed to some location). 
Reminds me of #16624

**Reply by thxkiwi** [reply]
With FetchContent, where do you see that fitting in to solve this? We're using Visual Studio (and therefore msbuild under the hood). The only other place we have CMake is the portfile for the port for Library.

It feels to me like disabling binary caching for a specific port (i.e. the port for Library) would allow the rebuild of Library - at least after removing the vcpkg_installed - while still utilizing the cached versions of other ports. 

To me, it isn't the switching between release and dev that's the problem (we can do that with a simple change of the vcpkg registry). It's that the previous dev build, regardless how or when it was obtained and built, has been cached. The caching definitely makes sense and I wouldn't want to change those semantics but more have finer control there over. 

Perhaps I'm thinking too low level though.

**Reply by thxkiwi** [reply]
Thinking higher level, it does like the high level use case has some intersection with #16624.

**Reply by autoantwort** [reply]
Would something like `--head=portname`, that only builds `portname` as head dependency and disables binary caching for that lib solve your issue? Or do you prefer when it do binary caching but includes the HEAD commit sha in the binary cache computation?

**Reply by thxkiwi** [reply]
@autoantwort I didn't see this originally. 

Disabling binary caching would be ideal because, in this case, the package is not built from source via vcpkg. The portfile only takes prebuilt artifacts and distributes them via vcpkg along with the binary libraries on which my Library depends.

What I've ended up hacking together is a script that uses the status yaml database to 
1. Remove the installed files from a vcpkg_installed directory
2. Remove the archive from the binary cache archives tree
3. Remove the record for the package from the status database

There is something related to the state of the vcpkg_installed tree that I don't understand at this point. Happy to go into details if anyone is interested but I'd rather have the ability disable binary caching.

I run this script as a pre-build event so that the App using Library is always linking Library based on HEAD since both App and Library are in the same repo.
